packages/interaction-tracking/src/InteractionTracking.js

```diff
@@ -0,0 +1,311 @@
+  onWorkScheduled: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of scheduled work has been cancelled.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkCancelled: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of work has started for a set of interactions.
+  // When this work is complete, onWorkStopped will be called.
+  // Work is not always completed synchronously; yielding may occur in between.
+  // A batch of async/yieldy work may also be re-started before completing.
+  // In that case, onWorkStarted may be called more than once before onWorkStopped.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkStarted: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of work has completed for a set of interactions.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkStopped: (interactions: Set<Interaction>, threadID: number) => void,
+};
+
+export type InteractionsRef = {
+  current: Set<Interaction>,
+};
+
+export type Subscribers = Set<Subscriber>;
+export type ScheduledAsyncWorkCounts = Map<Interaction, number>;
+
+const DEFAULT_THREAD_ID = 0;
+
+// Counters used to generate unique IDs.
+let interactionIDCounter: number = 0;
+let threadIDCounter: number = 0;
+
+// Set of currently tracked interactions.
+// Interactions "stack"–
+// Meaning that newly tracked interactions are appended to the previously active set.
+// When an interaction goes out of scope, the previous set (if any) is restored.
+let interactionsRef: InteractionsRef | null = null;
+
+// Listener(s) to notify when interactions begin and end.
+// Note that subscribers are only supported when enableInteractionTrackingObserver is enabled.
+let subscribers: Subscribers | null = null;
+
+// Tracks the number of async operations scheduled for each interaction.
+// Once the number of scheduled operations drops to 0,
+// Interaction subscribers will be notified that the interaction has ended.
+// Note that pending counts are only tracked when enableInteractionTrackingObserver is enabled.
+let scheduledAsyncWorkCounts: ScheduledAsyncWorkCounts | null = null;
+
+if (enableInteractionTracking) {
+  interactionsRef = {
+    current: new Set(),
+  };
+
+  if (enableInteractionTrackingObserver) {
+    subscribers = new Set();
+    scheduledAsyncWorkCounts = new Map();
+  }
+}
+
+// These values are exported for libraries with advanced use cases (i.e. React).
+// They should not typically be accessed directly.
+export {
+  interactionsRef as __interactionsRef,
+  subscribers as __subscribers,
+  scheduledAsyncWorkCounts as __scheduledAsyncWorkCounts,
+};
+
+export function getCurrent(): Set<Interaction> | null {
+  if (!enableInteractionTracking) {
+    return null;
+  } else {
+    return ((interactionsRef: any): InteractionsRef).current;
+  }
+}
+
+export function getThreadID(): number {
```

Comment:
How are you going to use this? I suspect it'll be expensive to store. We could just use the expiration time.

Reply:
Why would this be expensive to store? It's just a number.

You can see how I use it in PR #13253. Each root gets its own base thread ID, which gets combined with the expiration time to uniquely identify a root+priority. I thought this was part of what we agreed to yesterday? Maybe I misundestood.

Reply:
I see. I was confused because this ID won’t be directly passed to the observer. Maybe it should be called getBaseThreadID or something?

Reply:
Ha, fair enough. I named it "base thread id" initially but then changed my mind. I'll change it back.

Reply:
To be clear, the reason I changed it was because this _could_ be the actual thread ID. It's an implementation detail of React that a single root spawns multiple "threads" for different priorities. You could imagine something else (e.g. Preact) that only has a single "thread" just using this return value directly.

For that reason, I still slightly prefer it as-is.